pkg/virt-controller/watch/snapshot.go

```diff
@@ -0,0 +1,777 @@
+}
+
+func vmSnapshotError(vmSnapshot *vmssv1alpha1.VirtualMachineSnapshot) *vmssv1alpha1.VirtualMachineSnapshotError {
+	if vmSnapshot.Status != nil && vmSnapshot.Status.Error != nil {
+		return vmSnapshot.Status.Error
+	}
+	return nil
+}
+
+func vmSnapshotProgressing(vmSnapshot *vmssv1alpha1.VirtualMachineSnapshot) bool {
+	return vmSnapshotError(vmSnapshot) == nil &&
+		(vmSnapshot.Status == nil || vmSnapshot.Status.ReadyToUse == nil || *vmSnapshot.Status.ReadyToUse == false)
+}
+
+func getVMSnapshotContentName(vmSnapshot *vmssv1alpha1.VirtualMachineSnapshot) string {
+	if vmSnapshot.Status != nil && vmSnapshot.Status.VirtualMachineSnapshotContentName != nil {
+		return *vmSnapshot.Status.VirtualMachineSnapshotContentName
+	}
+
+	return fmt.Sprintf("%s-%s", "vmsnapshot-content", vmSnapshot.UID)
+}
+
+// variable so can be overridden in tests
+var currentTime = func() *metav1.Time {
+	t := metav1.Now()
+	return &t
+}
+
+func (ctrl *SnapshotController) updateVMSnapshot(vmSnapshot *vmssv1alpha1.VirtualMachineSnapshot) error {
+	log.Log.V(3).Infof("Updating VirtualMachineSnapshot %s/%s", vmSnapshot.Namespace, vmSnapshot.Name)
+
+	// Make sure status is initialized
+	if vmSnapshot.Status == nil {
+		return ctrl.updateSnapshotStatus(vmSnapshot)
+	}
+
+	source, err := ctrl.getSnapshotSource(vmSnapshot)
+	if err != nil {
+		return err
+	}
+
+	// unlock the source if done/error
+	if !vmSnapshotProgressing(vmSnapshot) && source != nil && source.Locked() {
+		if err = source.Unlock(); err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	// check deleted
+	if vmSnapshot.DeletionTimestamp != nil {
+		return ctrl.cleanupVMSnapshot(vmSnapshot)
+	}
+
+	if source != nil && vmSnapshotProgressing(vmSnapshot) {
+		// attempt to lock source
+		// if fails will attempt again when source is updated
+		if !source.Locked() {
+			locked, err := source.Lock()
+			if err != nil {
+				return err
+			}
+
+			log.Log.V(3).Infof("Attempt to lock source returned: %t", locked)
+
+			return nil
+		}
+
+		// add source finalizer and maybe other stuff
+		updated, err := ctrl.initVMSnapshot(vmSnapshot)
+		if updated || err != nil {
+			return err
+		}
+
+		content, err := ctrl.getContent(vmSnapshot)
+		if err != nil {
+			return err
+		}
+
+		// create content if does not exist
+		if content == nil {
+			return ctrl.createContent(vmSnapshot)
+		}
+	}
+
+	if err = ctrl.updateSnapshotStatus(vmSnapshot); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ctrl *SnapshotController) updateVMSnapshotContent(content *vmssv1alpha1.VirtualMachineSnapshotContent) error {
+	log.Log.V(3).Infof("Updating VirtualMachineSnapshotContent %s/%s", content.Namespace, content.Name)
+
+	var volueSnapshotStatus []vmssv1alpha1.VolumeSnapshotStatus
+	var deletedSnapshots, skippedSnapshots []string
+
+	currentlyReady := content.Status != nil && content.Status.ReadyToUse != nil && *content.Status.ReadyToUse
```

Comment:
this is a little different than how our other operators work. The status section here is actually influencing how the reconcile loop converges on the desired state, rather than the status existing simply to report the observed state. 

More commonly I'd expect this function to first detect the status of all the volume snapshots and create whatever ones don't exist. Then i'd expect the vmSnapshotContent's status to reflect the current observable state of all the volume snapshots.

so basically, I'd expect `currentlyReady` to reflect information we gather from volume snapshot informers as it relates to this VMSnapshotContent object, and not information already stored in the VMSnapshotContent object itself.

Does that make sense? The distinction seems kind of subtle but it's a fairly large change in expected behavior.

Reply:
I understand your point, and for the most part, what you describe is true.  The problem is that there is only a small window in time in which it is valid to create snapshots.  And that is before the content ever reaches "ready" or "error" state.  During this time, what you describe is true.  Once one of those states is reached, the source is unlocked and there are no more consistency guarantees so we should not take any more snapshots.

As a concrete example, let's say a snapshot is "ready" at t0.  Then an administrator deletes a VolumeSnapshot or a VolumeSnapshot goes into error state at t1.  We DO NOT want to create another VolumeSnapshot at t2.

Reply:
yep, makes perfect sense. prior states actually do matter for snapshots.